module/zfs/dsl_dir.c

```diff
@@ -161,13 +162,95 @@ dsl_dir_evict_async(void *dbu)
 	if (dsl_deadlist_is_open(&dd->dd_livelist))
 		dsl_dir_livelist_close(dd);
 
+	rrm_enter_write(&dp->dp_spa->spa_ratelimit_lock);
+	if (dd->dd_ratelimit_root == dd) {
+		vfs_ratelimit_free(dd->dd_ratelimit);
+		dd->dd_ratelimit = NULL;
+		dd->dd_ratelimit_root = NULL;
+		/*
+		 * We don't have to recurse down, because there are no children.
+		 * If there were any, they will have a hold on us and we
+		 * couldn't be evicted.
+		 */
+	}
+	rrm_exit(&dp->dp_spa->spa_ratelimit_lock, NULL);
+
 	dsl_prop_fini(dd);
 	cv_destroy(&dd->dd_activity_cv);
 	mutex_destroy(&dd->dd_activity_lock);
 	mutex_destroy(&dd->dd_lock);
 	kmem_free(dd, sizeof (dsl_dir_t));
 }
 
+static boolean_t
+dsl_dir_ratelimit_read_properties(dsl_dir_t *dd, uint64_t *limits)
+{
+	char *myname, *setpoint;
+	boolean_t isset;
+	int type;
+
+	myname = kmem_alloc(ZFS_MAX_DATASET_NAME_LEN, KM_SLEEP);
+	dsl_dir_name(dd, myname);
+	setpoint = kmem_alloc(MAXNAMELEN, KM_SLEEP);
+
+	isset = B_FALSE;
+	for (type = ZFS_RATELIMIT_FIRST; type <= ZFS_RATELIMIT_LAST; type++) {
+		const char *propname;
+		uint64_t limit;
+
+		propname = zfs_prop_to_name(vfs_ratelimit_type_to_prop(type));
+		if (dsl_prop_get_dd(dd, propname, 8, 1, &limit, setpoint,
+		    B_FALSE) != 0) {
+			/* Property doesn't exist or unable to read. */
+			continue;
+		}
+		if (strcmp(myname, setpoint) != 0) {
+			/* Property is not set here. */
+			continue;
+		}
+		if (limit == 0) {
+			/* Property is set to none, but we treat it as unset. */
+			continue;
+		}
+		limits[type] = limit;
+		isset = B_TRUE;
+	}
+
+	kmem_free(setpoint, MAXNAMELEN);
+	kmem_free(myname, ZFS_MAX_DATASET_NAME_LEN);
+
+	return (isset);
+}
+
+static void
+dsl_dir_ratelimit_read(dsl_dir_t *dd)
+{
+	uint64_t limits[ZFS_RATELIMIT_NTYPES] = {0};
+	boolean_t isset, needlock;
+
+	isset = dsl_dir_ratelimit_read_properties(dd, limits);
+	needlock = !RRM_WRITE_HELD(&dd->dd_pool->dp_spa->spa_ratelimit_lock);
+
+	if (needlock) {
+		rrm_enter_write(&dd->dd_pool->dp_spa->spa_ratelimit_lock);
+	}
+	if (isset) {
+		dd->dd_ratelimit = vfs_ratelimit_alloc(limits);
+		dd->dd_ratelimit_root = dd;
+	} else {
+		dd->dd_ratelimit = NULL;
+		if (dd->dd_parent == NULL) {
+			dd->dd_ratelimit_root = NULL;
```

Comment:
This code is written as initial initialization, not caring about freeing previous values.  But I see in dsl_dir_hold_obj() you call dsl_dir_ratelimit_read() after calling to dmu_buf_init_user(), which means the user buffer is already public.  I wonder if there is a sufficient protection against somebody else accessing the uninitialized fields, or setting them to something that will be leaked due to overwrite here.

PS: If we move dsl_dir_ratelimit_read() before dmu_buf_init_user(), I guess we may not need locking, but we'd need a cleanup in case we lose.